### ISSUE TITLE
[TestServer] HttpContext.Connection containing wrong/default values

### ISSUE THREAD

**Issue by peterwurzinger** [aspnet/Hosting#1135]
Hi!
I don't know if it's actually an issue, so I'll just describe the behavior I've discovered.

I'm using Microsoft.AspNetCore.Hosting.TestHost.TestServer for integration testing purposes for a custom authentication middleware/handler.
The configuration is as follows
```cs
_server = new TestServer(new WebHostBuilder()
                .UseStartup<Startup>());
//Startup only contains configuration stuff for my middleware

_client = _server.CreateClient();
_client.GetAsync("/");
//...
```

When executing this, the request hits my middleware and I'm accessing `HttpContext.Connection`, where all exposed properties are set to their default values (Local/RemotePort = 0, other properties are set to `null`).
Am I missing a configuration step or something? We cannot get around testing whether the middleware is checking the clients IP-Address or not.

I tried to debug that based on the sources of Microsoft.AspNetCore.Http(Abstractions), but I didn't get too far with that.

EDIT: I'm using the preview2-final - packages of Microsoft.AspNetCore.*, if that makes any difference.

**Comment by davidfowl** [comment]
What are you expecting those to be set to?

**Comment by peterwurzinger** [comment]
Well my expectation was, to give a real example, that the property RemoteIPAddress would contain an instance of IPAddress containing the client's IP - e.g in an integration test scenario 127.0.0.1 or ::1.
Maybe my expectation is wrong, but since ALL properties seem to be unset, the ConnectionInfo instance is unusable in every use case I can imagine atm.

**Comment by Tratcher** [comment]
It's not even loopback, it's in memory.
What we really need is an easy way to provide values. Most test apps don't care.

**Comment by peterwurzinger** [comment]
So the client's request doesn't take all the way through the server's request pipeline and will therefore not trigger all modules that would normally handle the request - did I get that right @Tratcher?

**Comment by davidfowl** [comment]
The client is fake, the request is fake, it's an in memory pipeline, there is no networking involved. It *does* however, run the entire server pipeline so everything gets invoked like normal, just without network IO.

**Comment by peterwurzinger** [comment]
Okay, and due to the fact that there is no networking involved such OSI 3/4 artifacts like IP addresses won't be available..?

**Comment by davidfowl** [comment]
@peterwurzinger correct. All of that to say, we *could* set the IPAddress to localhost and it wouldn't be incorrect. It's just that we don't today so it's null.

**Comment by muratg** [comment]
@jkotalik let's set the IP to 127.0.0.1 or `::1`? and port to 80. Ping @Tratcher if you want to get the context. :)

**Comment by peterwurzinger** [comment]
@muratg Is it maybe a good idea to let the user decide which values those properties will hold by making it configurable instead of setting them to a static value? I'm thinking of cases (especially our use-cases ofc) where one would need to differentiate between remote IPs - e.g. if the caller IP is on a whitelist or not. One would then be able to test this behavior via unit/integration-tests.

Regards
Peter

**Comment by davidfowl** [comment]
@peterwurzinger if you want it to be configurable then I think we'd just provide a callback so that you can modify the feature collection before requests get sent to the server.

**Comment by peterwurzinger** [comment]
@davidfowl That sounds like a good solution to me.

**Comment by Tratcher** [comment]
There's a more general problem here. HttpClient is convenient, but there's a limit to how much you can use it to manipulate arbitrary server structures like the Connection IPs. The Katana version directly exposed the [Invoke](https://github.com/aspnet/AspNetKatana/blob/7cb459795158afe491ae3e6ea9d90a41606fed69/src/Microsoft.Owin.Testing/TestServer.cs#L156) method so you could manually create the server structures and call it. I wouldn't do it quite the same way again, but I do think we need an API that lets you create a HttpContext in your test code and hand that directly to the server without going through a HttpClient abstraction layer. This would give us full fidelity to manipulate server structures in the tests.

**Comment by peterwurzinger** [comment]
Is there a direct need to manipulate the RemoteIP via HttpClient (for every single request, to pursue that thought)? Maybe I don't get it right, but the returned client-instance from `TestServer.CreateClient()` shouldn't bother about doing so. Instead the TestServer instance should be able to take care of this business..?

But as already mentioned, maybe I just didn't get that right.

**Comment by Tratcher** [comment]
If you want to test how your application handles different RemoteIP values, then you need to be able to control those values from your test code. There's not much point in always using the same hardcoded values.

**Comment by peterwurzinger** [comment]
Well, maybe I just use the library wrong then. I create an instance of TestServer every time a single test is invoked, use it for this test and dispose it at the end. My thinking was, that not reusing server instances will leave me with test logic only without having to deal with potential artifacts left from previous tests.

With xUnit this doesn't affect test performance as bad as it sounds like but it's surely not the fastest option.

**Comment by Tratcher** [comment]
That's fine, but I don't see how it relates to controlling the ip values.

**Comment by peterwurzinger** [comment]
Yeah I don't really understand how those properties are populated, I guess I need to do a deep-dive into the code. I just assumed that there would be a higher chance of being able to manipulate those values in the course of creating the TestServer instance instead while/before sending requests via the HttpClient.

With that said, being able to populate the values while setting up the server for a single test would be fine for me - maybe this explains my weird thoughts.

**Comment by Tratcher** [comment]
Ah, that makes more sense. We could specify them per server, but they're per request data so I'd rather not. Yes I recommend making a server instance per test, but our tests often send more than one request.

**Comment by peterwurzinger** [comment]
> but they're per request data 

I absoluetly agree with that.

So basically there is a need for a mechanism to specify those values per request without providing a separate HttpClient-implementation? Imho the common denominator in this case is the HTTP-protocol itself..?
It sounds a bit hacky to me, but what about defining a set of experimental HTTP-Headers to populate unprovidable values?
e.g. `X-ASPNETHOSTING-RemoteIP=...`

**Comment by Tratcher** [comment]
These are below the Http protocol, we're trying to mock the TCP connection itself. We may want to do something similar for client certificates. As we're trying to mimic the real environment I don't think we should resort to workarounds like x- headers. We need direct control of the server data structures to be able to fully mock every aspect of a request.

**Comment by jkotalik** [comment]
@Tratcher Let's talk offline about the context.

**Comment by jkotalik** [comment]
I can see two options on how to fix this:
- Expose more options on the test server (LocalIpAddress, port, etc.)
- Expose the FeatureCollection on the TestServer for modification before the HttpRequest is sent out.

**Comment by Tratcher** [comment]
Yeah, #816 seems like a better way to address this. There are too many possible things you'd want to modify.

**Comment by Tratcher** [comment]
Resolved via #1248
